csharp/ql/src/semmle/code/csharp/dataflow/FrameworkDataFlow.qll

```diff
@@ -0,0 +1,313 @@
+   * on whether the type of the `i`th parameter of `c` is a collection type.
+   */
+  SummaryInput argument(FrameworkCallable c, int i, ContentList inputStack) {
+    result = argument(i) and
+    exists(Parameter p |
+      p = c.getParameter(i) and
+      if isCollectionType(p.getType())
+      then inputStack = ContentList::element()
+      else inputStack = ContentList::empty()
+    )
+  }
+
+  /**
+   * Gets an input specification that specifies output from the delegate at
+   * argument `i` as the input.
+   */
+  SummaryInput delegate(int i) { result = TDelegateSummaryInput(i) }
+
+  /**
+   * Gets an input specification that specifies output from the delegate at
+   * argument `i` as the input.
+   *
+   * `c` must be a compatible callable, that is, a callable where the `i`th
+   * parameter is a delegate.
+   */
+  SummaryInput delegate(FrameworkCallable c, int i) {
+    result = delegate(i) and
+    hasDelegateArgumentPosition(c, i)
+  }
+}
+
+class SummaryOutput = Impl::SummaryOutput;
+
+/** Provides predicates for constructing flow-summary output specifications. */
+module SummaryOutput {
+  /**
+   * Gets an output specification that specifies the qualifier in a call as
+   * the output.
+   */
+  SummaryOutput qualifier() { result = TQualifierSummaryOutput() }
+
+  /**
+   * Gets an output specification that specifies the return value from a call as
+   * the output.
+   */
+  SummaryOutput return() { result = TReturnSummaryOutput() }
+
+  /**
+   * Gets an output specification that specifies the `i`th argument in a call as
+   * the output.
+   */
+  SummaryOutput argument(int i) { result = TArgumentSummaryOutput(i) }
+
+  /**
+   * Gets an output specification that specifies parameter `j` of the delegate at
+   * argument `i` as the output.
+   */
+  SummaryOutput delegate(FrameworkCallable callable, int i, int j) {
+    result = TDelegateSummaryOutput(i, j) and
+    hasDelegateArgumentPosition2(callable, i, j)
+  }
+}
+
+class FlowSource = Impl::FlowSource;
+
+/** Provides predicates for constructing flow-source specifications. */
+module FlowSource {
+  class Range = Impl::FlowSource::Range;
+
+  private class RemoteFlowSource extends Range {
+    RemoteFlowSource() { this = "remote" }
+  }
+
+  FlowSource remote() { result = any(RemoteFlowSource s).toFlowSource() }
+
+  private class LocalFlowSource extends Range {
+    LocalFlowSource() { this = "local" }
+  }
+
+  FlowSource local() { result = any(LocalFlowSource s).toFlowSource() }
+
+  private class StoredFlowSource extends Range {
+    StoredFlowSource() { this = "stored" }
+  }
+
+  FlowSource stored() { result = any(StoredFlowSource s).toFlowSource() }
+}
+
+class FlowSink = Impl::FlowSink;
+
+/** Provides predicates for constructing flow-sink specifications. */
+module FlowSink {
+  class Range = Impl::FlowSink::Range;
+
+  private class RemoteFlowSink extends Range {
+    RemoteFlowSink() { this = "remote" }
+  }
```

Comment:
What's the correspondence between `FlowSink::Range` and the query suite? Today we roughly have one query per sink type, right? One for HTML, one for SQL, and so on. That's because each sink type needs a separate qhelp file, and each needs its own set of sanitizers.

Is `FlowSink::Range` really a "flow label"? I'd think it should go together with a group of sanitizers, which is also why I'd think these classes should be public. But maybe I'm asking these questions because I don't understand what these classes will be used for.

Reply:
It is true that many of the `FlowSink`s will be one-to-one with queries, but not all (for example the remote flow sink is useful for all exposure queries). The reason that `RemoteFlowSink` is not public is that it uses the underlying string representation; the predicate `remote()` on the other hand is public. So `Range` is only meant to be used when defining new kinds of flow sources/sinks.

The idea is that we can use these sources/sinks to redefine for example `RemoteFlowSource`:

```ql
class RemoteFlowSource extends Node {
  RemoteFlowSource() { any(FrameworkDataFlow fdf).hasSink(this, FlowSink::remote()) }
}
```

Reply:
Perhaps `RemoteFlowSink` should be called `RemoteFlowSinkRange`. The only reason I use the range-setup is to reduce string pitfalls; one can, of course, still construct an invalid string, but it is much more cumbersome (first you need to extend `Range` and then call `toFlowSink()` to get a value that wraps the invalid string).

I also forgot to apply `isSubsetOf` above, so it should be
```ql
class RemoteFlowSource extends Node {
  RemoteFlowSource() {
    any(FrameworkDataFlow fdf).hasSink(this, any(FlowSink fs | fs.isSubsetOf*(FlowSink::remote()))
  }
}
```